### PR TITLE
Clip to source tile grid in ReprojDataTile

### LINKED ISSUE #15417: Option to crop GeoTIFF source tiles to tile grid extent not respected if reprojected
**Describe the bug**
This can be seen in the example for #15402 if non-zero nodata value is specified
image
and possibly with other example sources if they are reprojected.

**Expected behavior**
Without reprojection the misleadingly named `wrapX` option in GeoTIFF is used to crop the output to the tile grid extent on both the x and y axes regardless of the source projection being global.  This is unrelated to what reprojection expects `wrapX` to mean - whether the output will be repeated on adjacent worlds on the x axis when reprojected to a global projection.  While it would be possible for ReprojDataTile to use this to crop the input to reprojection, with a major release pending can this be renamed to something more appropriate, as it should be possible to both crop to the original local extent and repeat in a global projection.

### PR DESCRIPTION
**Pull request by mike-000**
Fixes #15417

Only fixes the rendering.  Adds a `clipExtent` option to reprojection but setting it still relies on a DataTile source option name intended for another purpose which conflicts when reprojecting to a global projection.

Adding a separate clip to tile grid extent option to all tile sources would be more understandable and would allow both that and `wrapX` to be used independently, and could also be used with the canvas renderer, but that is beyond scope of fixing the rendering of reprojected GeoTIFFs.

Also avoids Chrome/Edge `getImageData are faster with the willReadFrequently attribute` warnings by not reusing a read canvas after releasing it.

### CODE REVIEW

**Review by ahocevar — CHANGES_REQUESTED** [review]
The code looks good, but I'm failing to understand the rendering test. Could you either add comments to the rendering test code, explaining what it tests, or think of a simpler test?

File: src/ol/reproj/DataTile.js [hunk 1]
```diff
@@ -394,7 +403,6 @@ class ReprojDataTile extends DataTile {
         );
 
         releaseCanvas(context);
-        canvasPool.push(canvas);
```
Comment (ahocevar) [inline src/ol/reproj/DataTile.js:397]: The removal of this line will likely cause problems in mobile Safari.

File: src/ol/source/DataTile.js [hunk 2]
```diff
@@ -239,6 +239,7 @@ class DataTileSource extends TileSource {
         gutter: this.getGutterForProjection(sourceProj),
         getTileFunction: (z, x, y, pixelRatio) =>
           this.getTile(z, x, y, pixelRatio, sourceProj),
+        clipExtent: !this.getWrapX(),
```
Comment (ahocevar) [inline src/ol/source/DataTile.js:242]: You tried to explain it here and in #15417, but I'm still failing to understand what `wrapX` has to do with this.

Don't we always want to clip to the source extent? Why make it optional?

File: test/rendering/cases/webgl-data-tile-clip-extent/expected.png
Comment (ahocevar) [inline test/rendering/cases/webgl-data-tile-clip-extent/expected.png:1]: This second test does not make the first one easier to understand.

I won't be able to finish reviewing this pull request without understanding what it fixes or improves. But like I said above, I tend to think we always want to clip to the source extent.

**Review by ahocevar — CHANGES_REQUESTED** [review]

File: src/ol/source/DataTile.js (hunk 2 above)
Comment (mike-000) [reply src/ol/source/DataTile.js:242]: As described in #15424 I do not think it `wrapX` should have anything to do with this, that is a misuse of `wrapX` and it should be a separate option.

We do not clip (and never have) to the tile grid extend in the canvas tile renderer, the tile grid extent merely prevents tiles which are completely outside the extent being requested, it does not stop tiles partly in the extent being rendered in full.  Typical example would be a regional (at detail level) map using a standard EPSG:3857 tile grid origin.  But as in the example there is a requirement for that with GeoTIFF sources (and it should not depend on `wrapX`).

File: src/ol/reproj/DataTile.js (hunk 1 above)
Comment (mike-000) [reply src/ol/reproj/DataTile.js:397]: Hopefully not as `releaseCanvas` is reducing the size to a single.

**Comment by ahocevar** [comment]
I was talking about the source extent, not the source tile grid extent. Isn't that what you're aiming for? To avoid things getting reprojected that shouldn't be rendered?

**Comment by mike-000** [comment]
> I was talking about the source extent, not the source tile grid extent. Isn't that what you're aiming for? To avoid things getting reprojected that shouldn't be rendered?

No, the WebGL tile renderer (unlike the canvas tile renderer) clips to the source tile grid extent when `wrapX` is false.  That is currently not reflected in ReprojDataTile.  If reprojecting COGs to a global projection there is still a requirement to clip (and if https://openlayers.org/en/latest/examples/multiple-cogs.html used two COGs in Fiji `wrapX` would need to be true).

File: test/rendering/cases/webgl-data-tile-clip-extent/expected.png
Comment (mike-000) [reply test/rendering/cases/webgl-data-tile-clip-extent/expected.png:1]: Compare `webgl-data-tile-clip-extent` with `webgl-data-tile-no-wrap`.  If those were debug tiles using the canvas renderer the tiles would be rendered in full in both cases.

**Comment by ahocevar** [comment]
If the goal is to respect a source's extent when reprojecting tiles (see https://github.com/openlayers/openlayers/issues/15417#issuecomment-1918705437), then the source could clip to the reprojected source extent. That would be more straightforward than carrying the source extent around with each tile and changing the tile reprojection code. If the WebGL renderer is currently not able to clip to arbitrary geometries (which a reprojected extent could become), then that would be a sensible addition to the WebGL renderer.

**Comment by mike-000** [comment]
No, the goal is to reproject exactly what is renderered when not reprojected - clip in the stitchcontext before creating the reproj tile.

**Comment by ahocevar** [comment]
But isn't the important thing to *render* the area of the extent that would be rendered when not reprojected? For that goal, it should not matter whether you clip for reprojection or for rendering.

I'm just trying to think of a solution that does not have to deal with whether or not the current interpretation of `wrapX` in the WebGL renderer makes sense or not.

**Comment by mike-000** [comment]
Borrowed from the raster reprojection example but the image is also available as a COG
image
It would be MUCH easier to clip that to a source extent before reprojection than after.

I think `wrapX` should be only be used to mean "Whether to wrap the world horizontally" as it does everywhere except GeoTIFF.  With a major release pending there should be a new property (e.g. `clipExtent`) for tile sources which is configurable for DataTileSource/GeoTIFF (defaulting to true) and set false for other sources (unless we also want to support it in the canvas renderer and reproj image tiles).

With `wrapX: true` that should allow adjacent Umbra type COGs from UTM zones 1 and 60 to been seen together in a global projection.

**Comment by ahocevar** [comment]
@mike-000 Again, unfortunately I have no idea what you're referring to. Why would users want to decide on whether to set a `clipExtent` property to `true` or `false`? As a user, I wouldn't even know which extent. And honestly, as a maintainer of this library who knows the code base quite well, I don't either.

So: why do we need a `clipExtent` option? Which extent do we clip?

**Comment by mike-000** [comment]
I would replace the existing `wrapX` in DataTileSource and GeoTIFF with
```
wrapX		boolean (defaults to false)	Whether to wrap the world horizontally.
clipExtent	boolean (defaults to true)	Do not render tiles beyond the tile grid extent.
```
i.e. give `wrapX` its normal meaning and replace the old non-standard use with a new option

Then change `!source.getWrapX()` to `source.getClipExtent()` in https://github.com/openlayers/openlayers/blob/main/src/ol/renderer/webgl/TileLayerBase.js#L117-L123

**Comment by ahocevar** [comment]
Agreed on the removal of the misuse of `wrapX` property in the renderer. But when would someone want to set `clipExtent` to `false`?

**Comment by mike-000** [comment]
For GeoTIFF unlikely (although the current misused `wrapX` can be changed), but for a DataTileSource we might want to be consistent with image tile sources and/or the  canvas renderer.

**Comment by ahocevar** [comment]
Always clipping to the tile grid extent would make the webgl-mixed-layers and webgl-multiple-layers rendering tests break. To fix that, the WebGL TileLayer renderer would have to repeat tiles when `wrapX` is true and the projection is a global projection, like the Canvas TileLayer renderer does.

So instead of introducing a `clipExtent` option, clipping to the tile grid extent should always happen when `wrapX` is false, and when `wrapX` is true and the projection is a global projection, tiles should be repeated like in the Canvas TileLayer renderer.

Another missing piece is to mask out the layer extent, as @tschaub suggested in https://github.com/openlayers/openlayers/pull/13060#issue-1066376808. That would bring consistency with image tile sources rendered by the canvas renderer.

**Comment by mike-000** [comment]
Yes, I know DataTile and GeoTIFF do not wrap as expected unless reprojected, the now very out of date #13727 attempted to fix that but did not consider the existing non-standard use of `wrapX`.

For reprojection (the scope of this pull request) I think all that is needed is for `ReprojDataTile` to unconditionally pass `clipExtent` true to `renderPreprojected`?

File: src/ol/reproj/DataTile.js (hunk 1 above)
Comment (ahocevar) [reply src/ol/reproj/DataTile.js:397]: Yes, but it takes a while for Safari to garbage collect that, so it is important to reuse the canvas.

**Comment by ahocevar** [comment]
> For reprojection (the scope of this pull request) I think all that is needed is for ReprojDataTile to unconditionally pass clipExtent true to renderPreprojected?

Yes, and I think that'll be it, with no further changes needed. For 99% of all use cases, the current use of `wrapX()` should work.

**Review by ahocevar — APPROVED** [review]
Thanks, @mike-000

**Pull request merged**